test/tools/ossfuzz/protomutators/YulProtoMutator.cpp

```diff
@@ -46,6 +46,34 @@ static LPMPostProcessor<Block> addStoreToZero(
 	}
 );
 
+
+namespace
+{
+template <typename T>
+struct addControlFlow
+{
+	addControlFlow()
+	{
+		function = [](T* _message, unsigned)
+		{
+			MutationInfo m{_message, "Added control flow."};
+			YPM::addControlFlow(_message);
+		};
+		/// Unused variable registers callback.
+		LPMPostProcessor<T> callback(function);
+	}
+	function<void(T*, unsigned)> function;
+};
+}
+
+static addControlFlow<ForStmt> c1;
+static addControlFlow<BoundedForStmt> c2;
+static addControlFlow<IfStmt> c3;
+static addControlFlow<SwitchStmt> c4;
+static addControlFlow<FunctionDef> c5;
+static addControlFlow<CaseStmt> c6;
+static addControlFlow<Code> c7;
```

Comment:
These are unused currently. How will you use these?

Reply:
The constructor of these static variables will "register" the callbacks (i.e., custom mutation routines) with libprotobuf-mutator. The registration happens here:

https://github.com/ethereum/solidity/pull/10967/files#diff-ec877dd39b51448b8cb75a828d9379e511d92ab334fac2ca38928cd7e8a9ed39R63

Reply:
Thanks. Could you explain a bit on how this callback works?

Reply:
iiuc, here is how the callback gets registered:

1. Entry point is this struct that lives in libprotobuf mutator (https://github.com/google/libprotobuf-mutator/blob/e5869dd9690c3f4dfb842fb90bd07a5a9ee32172/src/libfuzzer/libfuzzer_macro.h#L114-L124)
2. This entry point struct is aliased here: https://github.com/ethereum/solidity/blob/e75e3fc2e321cd00cabfe2cd0ab4b191eefd76ce/test/tools/ossfuzz/protomutators/YulProtoMutator.h#L16-L17
3. The entry point type is instantiated via a local variable inside the constructor of `addControlFlow` here: https://github.com/ethereum/solidity/pull/10967/files#diff-ec877dd39b51448b8cb75a828d9379e511d92ab334fac2ca38928cd7e8a9ed39R63
3.1 The instantiation of the local variable should call the constructor of the callback registration interface which is defined here: https://github.com/google/libprotobuf-mutator/blob/8942a9ba43d8bb196230c321d46d6a137957a719/src/libfuzzer/libfuzzer_macro.cc#L234-L239
3.2 That should eventually call this https://github.com/google/libprotobuf-mutator/blob/8942a9ba43d8bb196230c321d46d6a137957a719/src/mutator.cc#L683-L686
3.3 That should insert the passed callback into LPM's custom mutation map (see https://github.com/google/libprotobuf-mutator/blob/8942a9ba43d8bb196230c321d46d6a137957a719/src/mutator.h#L104-L106 and https://github.com/google/libprotobuf-mutator/blob/8942a9ba43d8bb196230c321d46d6a137957a719/src/mutator.h#L68-L69)

Once the map insertion happens, we are done registering our custom mutation with the libprotobuf mutator library. Any time, the library is then mutating a message whose description matches that against which a user-defined function was previously registered, the library is going to invoke it mutating the message how we want in the process.

I think the actual invocation of user-defined mutations for a message type happens here: https://github.com/google/libprotobuf-mutator/blob/8942a9ba43d8bb196230c321d46d6a137957a719/src/mutator.cc#L494-L497

I think the whole sequence requires the use of static variables so that the one-time "registration" happens early. See https://github.com/google/libprotobuf-mutator/blob/e5869dd9690c3f4dfb842fb90bd07a5a9ee32172/src/libfuzzer/libfuzzer_macro.h#L37-L40

Reply:
Thanks